Automattic/jetpack
# Docker environment: Introduce several changes now that the base work is in master.

This is a follow-up to #8949. We merged #8949 to have a basis and be able to start working/developing using the initial docker setup.

Refer to [docker/README.md](https://github.com/Automattic/jetpack/blob/update/flyby-changes-to-docker-scripts/docker/README.md) for details. (Also don't forget to update it in this branch if it makes sense).

This PR should be used for dropping flyby commits that improve the setup. Also it's meant to receive commits by anyone who wants to make the setup work on environments like Windows, WSL, linux if it currently fails. We mostly developed this on MacOS, so the original work is certainly biased.

pinging @enejb @simison @roccotripaldi @lezama @dereksmart @gititon @sirreal @withinboredom .

#### Changes proposed in this Pull Request:

Introduces flyby changes:

* Ability to detect https for `WP_SITEURL` and `WP_HOME` if the request is being forwarded by a reverse proxy using SSL. (Like ngrok does).
* Adds SFTP container for easier Rewind testing
* Makes `yarn docker:*` commands more Windows friendly and less noisy
* Disables ability to delete Jetpack plugin via wp-admin since this would wipe out the whole development directory
* Other small general improvements

#### TODO/ideas
- [x] Configure https to work via Ngrok
- [ ] Publish Docker image to registry and use it from there
- [ ] Multiple PHP versions for running phpunit
mac host environments
- [ ] env variables for easy config (php version, WP version...)
- [ ] Add [Query monitor](https://wordpress.org/plugins/query-monitor/) plugin
- [ ] Easy filling with massive amount of stuff ([wptest](https://github.com/poststatus/wptest) is good for that. Comes with a WP-CLI script too.)
- [ ] Add memcached?
- [ ] re-install WP without deleting containers/images?
- [ ] Use bash & bat scripts to operate instead of yarn/npm scripts
- [ ] local https certs?
- [x] [sFTP](https://hub.docker.com/r/atmoz/sftp/) container (to test Rewind)
- [ ] Xdebug: Currently, we're missing a way to have the host IP addressed be referenced successfully in non-mac host environments
- [ ] Xdebug instructions for PHPStorm
- [ ] Fix warning: `apache2: Could not reliably determine the server’s fully qualified domain name, using 172.22.0.5. Set the ‘ServerName’ directive globally to suppress this message` (`extra_hosts` in `docker-compose.yml` was an attempt to fix this but IPs are dynamic)
- [ ] Lint files (came up by modifying `mu-plugins/*`) ?
- [ ] Use more stable version of WP-CLI. Now `2.0.0-alpha`


#### Testing instructions:

Whenever you get to this branch, make sure you do the following to test the changes:

* `yarn docker:stop` (Make sure you stop all running containers related to the setup).
* `yarn docker:build` (Make sure you get to rebuild the `wordpress` image if changes were introduced in `Dockerfile` and also copies the template files again in case they were modified).
* `yarn docker:up`. Start the containers.

Also, remember to test if the following still works:

* Accessing the site via `http://localhost`.
* Accessing the site via `http://some-reverse-proxy` like ngrok.
* Accessing the site via `https://some-reverse-proxy` like ngrok.
* Running `yarn docker:phpunit --filter=Protect` to see if tests still run ok.
